### PR TITLE
Re-implement using callbacks.

### PR DESCRIPTION
**Pull request by quelgar**
Re-implement using callbacks.

Advantages:

* The ZIO → Monix conversion is now lazy
* Interruption/cancellation can be propagated
* Executes synchronously where possible

Also removed the need for the user to provide a Monix scheduler, instead we create one from the ZIO runtime.

### CODE REVIEW

File: interop-monix/shared/src/main/scala/zio/interop/monix/package.scala [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      Task.runtime.flatMap { zioRuntime =>
+        Task.effectAsyncInterrupt[A] { cb =>
+          implicit val scheduler: MScheduler = zioRuntime.monixScheduler(executionModel)
+          try
+          // runSyncStep will try to execute the Monix effects synchronously
+          // if it fails before hitting an async boundary, the failure will be thrown
+          monixTask.runSyncStep match {
+            case Right(result)        =>
+              // Monix task ran synchronously and successfully
+              Right(ZIO.succeed(result))
```
Comment (mijicd) [inline interop-monix/shared/src/main/scala/zio/interop/monix/package.scala:55]:
```suggestion
              Right(ZIO.succeedNow(result))
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        testM("converts a synchronous Monix task to a ZIO task") {
+          @volatile var testVar = 0
+          val monixTask         = MTask.eval {
+            val old = testVar
+            testVar = old + 1
+            old
+          }
+          val io                = for {
+            orig    <- ZIO.fromMonixTask(monixTask)
+            current <- ZIO.succeed(testVar)
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:40]:
```suggestion
            current <- UIO(testVar)
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        testM("converts an asynchronous Monix task to a ZIO task") {
+          @volatile var testVar = 0
+          val monixTask         = MTask.async[Int] { cb =>
+            val old = testVar
+            testVar = old + 1
+            cb.onSuccess(old)
+          }
+          val io                = for {
+            orig    <- ZIO.fromMonixTask(monixTask)
+            current <- ZIO.succeed(testVar)
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:53]:
```suggestion
            current <- UIO(testVar)
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        testM("lazily executes a converted Monix task") {
+          @volatile var executed = false
+          val monixTask          = MTask.eval {
+            executed = true
+          }
+          val test               = ZIO.succeed {
+            executed = false
+          } *> {
+            ZIO.fromMonixTask(monixTask)
+            ZIO.succeed(executed)
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:74]: Should these two be chained? Also, `ZIO.succeed` => `UIO`.

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            }
+          }
+          for {
+            _            <- ZIO.succeed {
+                              cancelled = false
+                              running = false
+                            }
+            fiber        <- ZIO.fromMonixTask(monixTask).fork
+            // wait until the monix task is running before interrupting the fiber
+            _            <- ZIO.succeed(running).repeatUntil(Predef.identity)
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:95]:
```suggestion
            _            <- UIO(running).repeatUntil(Predef.identity)
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            exit         <- fiber.interrupt
+            wasCancelled <- ZIO.succeed(cancelled)
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:97]:
```suggestion
            wasCancelled <- UIO(cancelled)
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+          @volatile var running   = false
+          val monixTask           = MTask.eval {
+            running = true
+          } >> MTask.never.doOnCancel {
+            MTask.eval {
+              cancelled = true
+            }
+          }
+          for {
+            _            <- ZIO.succeed {
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:89]:
```suggestion
            _            <- ZIO.effectTotal {
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+          }
+          val io        = ZIO.fromMonixTask(monixTask)
+          assertM(io.either)(isLeft(equalTo(error)))
+        },
+        testM("lazily executes a converted Monix task") {
+          @volatile var executed = false
+          val monixTask          = MTask.eval {
+            executed = true
+          }
+          val test               = ZIO.succeed {
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:70]:
```suggestion
          val test               = ZIO.effectTotal {
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            _            <- ZIO.succeed(running).repeatUntil(Predef.identity)
+            exit         <- fiber.interrupt
+            wasCancelled <- ZIO.succeed(cancelled)
+          } yield assertTrue(wasCancelled) && assertTrue(exit.interrupted)
+        } @@ nonFlaky
+      ),
+      suite("ZIO#toMonixTask")(
+        testM("converts a successful ZIO task to a Monix task") {
+          @volatile var testVar = 0
+          val io                = ZIO.succeed {
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:104]:
```suggestion
          val io                = ZIO.effectTotal {
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        testM("converts a failed ZIO task to a failed Monix task") {
+          val error = new Exception("ZIO operation failed")
+          val io    = ZIO.fail(error)
+          val test  = io.toMonixTask.flatMap[Any, Throwable, Nothing](ZIO.fromMonixTask(_)).either
+          assertM(test)(isLeft(equalTo(error)))
+        },
+        testM("propagates cancellation from Monix to ZIO") {
+          @volatile var cancelled = false
+          @volatile var running   = false
+          val io                  = ZIO.succeed {
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:124]:
```suggestion
          val io                  = ZIO.effectTotal {
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+            running = true
+          } *> ZIO.never.onInterrupt {
+            ZIO.succeed {
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:127]:
```suggestion
            ZIO.effectTotal {
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+            _            <- ZIO.fromMonixTask {
+                              for {
+                                fiber <- monixTask.start
+                                _     <- MTask.eval(running).restartUntil(Predef.identity)
+                                _     <- fiber.cancel
+                              } yield ()
+                            }
+            // Monix `fiber.cancel` doesn't wait for all the cancellation effects
+            // to complete, so `cancelled` might not be true at first
+            wasCancelled <- ZIO.succeed(cancelled).repeatUntil(Predef.identity)
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:142]:
```suggestion
            wasCancelled <- UIO(cancelled).repeatUntil(Predef.identity)
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+          } yield wasCancelled
+          assertM(test)(isTrue)
+        } @@ jvmOnly @@ timeout(5.seconds),
+        testM("only executes the ZIO effect if the Monix task is executed") {
+          @volatile var executed: Boolean = false
+          val io                          = ZIO.succeed {
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:148]:
```suggestion
          val io                          = ZIO.effectTotal {
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+            executed = true
+          }
+          val test                        = for {
+            _      <- ZIO.succeed {
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:152]:
```suggestion
            _      <- ZIO.effectTotal {
```

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+                        executed = false
+                      }
+            _      <- io.toMonixTask
+            result <- ZIO.succeed(executed)
```
Comment (mijicd) [inline interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:156]:
```suggestion
            result <- UIO(executed)
```

**Review by mijicd — COMMENTED** [review]
Please add some spaces to improve formatting. Other than that and a few nitpicks, this looks good to me!

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala (hunk 4 above)
Comment (quelgar) [reply interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:74]: I'm trying to prove that the Monix task isn't executed just by creating the ZIO value wrapping it, we'd have to actually run the ZIO effect. I couldn't think of a better way.

File: interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala (hunk 11 above)
Comment (quelgar) [reply interop-monix/shared/src/test/scala/zio/interop/monix/MonixTaskSpec.scala:127]: I somehow got the idea that in ZIO 2, `succeed` was preferred.

**Review by mijicd — APPROVED** [review]

**Review by mijicd — APPROVED** [review]

**Pull request merged**